interfaces/policy/policy.go

```diff
@@ -0,0 +1,111 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package policy implements the declaration based policy checks for
+// connecting or permitting installation of snaps based on their slots
+// and plugs.
+package policy
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/snap"
+)
+
+// TODO: InstallCandidate
+
+// ConnectCandidate represents a candidate connection.
+type ConnectCandidate struct {
+	// TODO: later we need to carry dynamic attributes once we have those
+	Plug                *snap.PlugInfo
+	PlugSnapDeclaration *asserts.SnapDeclaration
+
+	Slot                *snap.SlotInfo
+	SlotSnapDeclaration *asserts.SnapDeclaration
+
+	BaseDeclaration *asserts.BaseDeclaration
+}
+
+func (connc *ConnectCandidate) plugAttrs() map[string]interface{} {
+	return connc.Plug.Attrs
+}
+
+func (connc *ConnectCandidate) slotAttrs() map[string]interface{} {
+	return connc.Slot.Attrs
+}
+
+func (connc *ConnectCandidate) plugSnapType() snap.Type {
+	return connc.Plug.Snap.Type
+}
+
+func (connc *ConnectCandidate) slotSnapType() snap.Type {
+	return connc.Slot.Snap.Type
+}
+
+func (connc *ConnectCandidate) checkPlugRule(rule *asserts.PlugRule, whichDecl string) error {
+	if checkPlugConnectionConstraints(connc, rule.DenyConnection) == nil {
+		return fmt.Errorf("connection denied because it matches deny-connection in plug rule for interface %q from %s", connc.Plug.Interface, whichDecl)
```

Comment:
Sounds like too many details for normal people. Perhaps we can be subtle so that we can distinguish, but without being over-verbose. Perhaps something along the lines of:
```
// deny-connection
=> connection denied by plug rule of interface "network" for "core" snap
// allow-connection
=> connection not allowed by plug rule of interface "network" for "core" snap"
```
For base-declaration, we might just omit the snap name:
```
// deny-connection
=> connection denied by plug rule of interface "network"
// allow-connection
=> connection not allowed by plug rule of interface "network"
```
Similarly for slots below.